README.md
# AppDynamics Cookbook

This is a fork from [appdynamics-cookbooks](https://github.com/Appdynamics/appdynamics-cookbooks) reduced to the Appdynamics machine and java agent with some fixes and configuration extends. It is fully productive used in a well known insurance company in Switzerland for Tomcat nodes.

Learn more about AppDynamics at:

* http://www.appdynamics.com/ (and check out the handsome devil next to the "Act" bubble in the photo)
* https://docs.appdynamics.com/display/PRO41/Getting+Started
* https://docs.appdynamics.com/display/PRO41/Install+and+Administer+Agents

## Requirements

* Chef >= 0.10.0
* unzip

## Attributes

### Default Attributes

These node attributes must be set to use the `_agent` recipes:

* `node['appdynamics']['app_name']` - The name to register your application under with the AppDynamics controller.
* `node['appdynamics']['tier_name']` - The name to register this tier of your application under with the AppDynamics controller.
* `node['appdynamics']['node_name']` - The name to register this node of your application under with the AppDynamics controller. Default is the `host name`
* `node['appdynamics']['unique_host_id_name']` - The name to register the unique host id name of your application under with the AppDynamics controller. Default is the `host name`
* `node['appdynamics']['version']` - The version of AppDynamics to use. Agent version overrides this. The package URI will be composed with the version
* `node['appdynamics']['controller']['host']` - The host your AppDynamics controller is running on (a domain name or IP address).
* `node['appdynamics']['controller']['port']` - The port your AppDynamics controller is running on.
* `node['appdynamics']['controller']['user']` - The account name to use with your AppDynamics controller.
* `node['appdynamics']['controller']['accesskey']` - The access key for your account for accessing your AppDynamics controller.

Optional attributes:

* `node['appdynamics']['packages_site']` - The base URL of the AppDynamics packages site (defaults to `https://packages.appdynamics.com`).
* `node['appdynamics']['controller']['ssl']` - Flag indicating if SSL should be used to speak to the controller (`true`) or not (`false`). Defaults to `true`. SaaS controllers do not support the value `false` for this flag.

### HTTP Proxy Attributes

If your agents must use an HTTP proxy to communicate with the controller, set these attributes:

* `node['appdynamics']['http_proxy']['host']` - The hostname/IP of your HTTP proxy.
* `node['appdynamics']['http_proxy']['port']` - The port of your HTTP proxy.
* `node['appdynamics']['http_proxy']['user']` - If you must authenticate with your HTTP proxy, the user name to authenticate as.
* `node['appdynamics']['http_proxy']['password_file']` - If you must authenticate with your HTTP proxy, the complete path to a file you will create on a machine using an AppDynamics agent that contains the password used for authenticating with your HTTP proxy.

#### Machine Agent Configuration Attributes

The `machine_agent` recipe has some additional attributes you may set:

* `node['appdynamics']['machine_agent']['version']` - The version of the AppDynamics package for the machine agent
* `node['appdynamics']['machine_agent']['source']` - Set complete URI of zip package to download from. Default `nil` because it is composed by package_site, version, (machineagent-bundle-, arch, OS) AgentName and .zip suffix.
* `node['appdynamics']['machine_agent']['use_bundled_package']` - Include machineagent-bundle-, arch and OS into source URI
* `node['appdynamics']['machine_agent']['checksum']` - The sha256 checksum of the source package
* `node['appdynamics']['machine_agent']['install_dir']` - Installation directory of the machine agent. Default `/opt/appdynamics/machineagent`
* `node['appdynamics']['machine_agent']['owner']` - User of the machine agent. Default `root`
* `node['appdynamics']['machine_agent']['group']` - Group of the machine agent. Default `root`
* `node['appdynamics']['machine_agent']['init_script']` - Path and name of the init script. Default `/etc/init.d/appdynamics_machine_agent`
* `node['appdynamics']['machine_agent']['pid_file']` - Path and name of the pid file. Default `/var/run/appdynamics_machine_agent.pid`
* `node['appdynamics']['machine_agent']['template']['cookbook']` - Name of the cookbook template folder for the controller-info-xml. Default `appdynamics`
* `node['appdynamics']['machine_agent']['template']['source']` - Location of the template file for the controller-info.xml . Default `machine/controller-info.xml.erb`
* `node['appdynamics']['machine_agent']['init']['cookbook']` - Name of the cookbook template folder for the init script. Default `appdynamics`
* `node['appdynamics']['machine_agent']['init']['source']` - Location of the template file for the init script. Default `machine/init.d.erb`
* `node['appdynamics']['machine_agent']['run_sh']['cookbook']` - Name of the cookbook template folder for the run.sh script. Default `appdynamics`
* `node['appdynamics']['machine_agent']['run_sh']['source']` - Location of the template file for the run.sh script. Default `machine/run.sh.erb`
* `node['appdynamics']['machine_agent']['java']` - Path of the java binary. Default `/usr/bin/java`
* `node['appdynamics']['machine_agent']['java_params']` - System properties to pass to the JVM of the machine agent (space separated). Default `Xmx32m`

#### Java Agent Configuration Attributes

* `node['appdynamics']['java_agent']['version']` - The version of the AppDynamics package for the java agent
* `node['appdynamics']['java_agent']['source']` - Set complete URI of zip package to download from. Default `nil` becaus
e it is composed by package_site, /java/, version, AppServerAgent-, (ibm-), version and .zip suffix.
* `node['appdynamics']['java_agent']['ibm_jvm']` - Include "ibm-" between AppServerAgent- and version.zip in the composed package URI. Default `false`
* `node['appdynamics']['java_agent']['checksum']` - The sha256 checksum of the source package
* `node['appdynamics']['java_agent']['install_dir']` - Installation directory of the java agent. The package contains the root folder "javaagent". Default `/opt/appdynamics`
* `node['appdynamics']['java_agent']['owner']` - User of the java agent. To avoid permission problem, set the same user as your JVM is running (tomcat, jboss, etc) . Default `root`
* `node['appdynamics']['java_agent']['group']` - Group of the java agent. Default `root`

* `node['appdynamics']['java_agent']['template']['cookbook']` - Name of the cookbook template folder for the controller-info-xml. This is normally not needed as you pass java agent settings through system properties of your JVM. Default `appdynamics`
* `node['appdynamics']['java_agent']['template']['source']` - Location of the template file for the controller-info.xml. his is normally not needed as you pass java agent settings through system properties of your JVM. Default `machine/controller-info.xml.erb`
* `node['appdynamics']['java_agent']['zip']` - The file name of the package to be saved on the file system. Default `#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip`

## Example usage

**Step 1.** Define attributes.

```ruby
default_attributes (
  'appdynamics' => {
    'machine_agent' => {
      'version: => '4.5.7.1975'
    },
    'java_agent' => {
      'version' => '4.5.6.24621',
      'owner' => 'tomcat',
      'group' => 'tomcat'
    },
    'app_name' => 'my app',
    'tier_name' => 'frontend',
    'node_name' => node.name,
    'unique_host_id_name' => node.name,
    'controller' => {
      'host' => 'my-controller',
      'port' => '8081',
      'ssl' => false,
      'user' => 'someuser',
      'accesskey' => 'supersecret',
    }
  }
)
```

**Step 2.** Add `recipe[appdynamics::machine_agent]` and/or `recipe[appdynamics:java_agent]` to your run list.

## Built With

* [chef](https://www.chef.io) - Chef Software, Inc

## Contributing

Please read [CONTRIBUTING](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/your/project/tags). 

## Authors

* **Roland Hansmann** - *Initial work* - [Roland Hansmann](https://github.com/rediculum)

See also the list of [contributors](https://github.com/avectris/appdynamics/graphs/contributors) who participated in this project.

## License

This project is licensed under the GNU Affero General Public License v3.0 License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# appdynamics
========================

This file is used to list changes made in each version of the appdynamics cookbook.

## v1.0.0 (06.02.2019)

- Official release on github

## v0.2.5 (04.02.2019)

- owner of logs directory from java agent by specified attribute

## v0.2.4 (22.01.2019)

- remove unique-host-id from machine agent template

## v0.2.3 (16.01.2019)

- restart agent on controller-info.xml changes

## v0.2.2 (16.01.2019)

- add unique-host-id element into controller-info.xml

## v0.2.1 (19.12.2018)

- set correct ownership to conf directory of java_agent

## v0.2.0 (18.12.2018)

- add java agent
- fix package extraction handling for machine agent

## v0.1.1 (18.12.2018)

- remove old lsb function sourcing in init script

## v0.1.0 (10.12.2018)

- reduced original appdynamics cookbook to only handle the machine agent


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  config.log_level = :error
  config.color = true
  config.formatter = :documentation
end


test/integration/default/java_agent_test.rb
%w(conf logs).each do |dir|
  describe file(['/opt/appdynamics/javaagent/', dir].join) do
    it { should be_directory }
    it { should exist }
    its('mode') { should cmp '0755' }
    its('owner') { should eq 'vagrant' }
    its('group') { should eq 'vagrant' }
  end
end

describe file('/opt/appdynamics/javaagent/javaagent.jar') do
  it { should exist }
end

%w(external-services lib sdk sdk-plugins utils).each do |dir|
  describe file(['/opt/appdynamics/javaagent/', dir].join) do
    it { should be_directory }
    it { should exist }
  end
end

describe file('/opt/appdynamics/javaagent/conf/controller-info.xml') do
  its('mode') { should cmp '0600' }
  its('owner') { should eq 'vagrant' }
  its('group') { should eq 'vagrant' }
  its('content') { should match(%r{controller-port>443<\/controller-port>$}) }
  its('content') { should match(%r{controller-host>localhost<\/controller-host>$}) }
  its('content') { should match(%r{controller-ssl-enabled>true<\/controller-ssl-enabled>$}) }
  its('content') { should match(%r{account-name>controller-user<\/account-name>$}) }
  its('content') { should match(%r{account-access-key>ABCDEF1234567890<\/account-access-key>$}) }
  its('content') { should match(%r{application-name>tomcat<\/application-name>$}) }
  its('content') { should match(%r{tier-name>development<\/tier-name>$}) }
  its('content') { should match(%r{node-name>test-node<\/node-name>$}) }
end


test/integration/default/machine_agent_test.rb
describe package('unzip') do
  it { should be_installed }
end

describe file('/opt/appdynamics/machineagent/conf') do
  it { should be_directory }
  it { should exist }
  its('mode') { should cmp '0755' }
  its('owner') { should eq 'root' }
  its('group') { should eq 'root' }
end

describe file('/opt/appdynamics/machineagent/run.sh') do
  it { should exist }
  its('mode') { should cmp '0744' }
  its('owner') { should eq 'root' }
  its('group') { should eq 'root' }
end

describe file('/opt/appdynamics/machineagent/machineagent.jar') do
  it { should exist }
end

%w(bin extensions lib local-scripts monitors monitorsLibs scripts).each do |dir|
  describe file(['/opt/appdynamics/machineagent/', dir].join) do
    it { should be_directory }
    it { should exist }
  end
end

describe file('/etc/init.d/appdynamics_machine_agent') do
  it { should exist }
  its('mode') { should cmp '0744' }
  its('owner') { should eq 'root' }
  its('group') { should eq 'root' }
  its('content') { should match(%r{^INSTALL_DIR="/opt/appdynamics/machineagent"$}) }
  its('content') { should match(%r{^pid_file="/var/run/appdynamics_machine_agent.pid"$}) }
end

describe file('/opt/appdynamics/machineagent/conf/controller-info.xml') do
  its('mode') { should cmp '0600' }
  its('owner') { should eq 'root' }
  its('group') { should eq 'root' }
  its('content') { should match(%r{controller-host>localhost<\/controller-host>$}) }
  its('content') { should match(%r{account-name>controller-user<\/account-name>$}) }
  its('content') { should match(%r{account-access-key>ABCDEF1234567890<\/account-access-key>$}) }
  its('content') { should match(%r{application-name>tomcat<\/application-name>$}) }
  its('content') { should match(%r{tier-name>development<\/tier-name>$}) }
  its('content') { should match(%r{node-name>test-node<\/node-name>$}) }
end

describe service('appdynamics_machine_agent') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end


recipes/java_agent.rb
include_recipe 'ark'

agent = node['appdynamics']['java_agent']
controller = node['appdynamics']['controller']

version = agent['version'] || node['appdynamics']['version']
raise 'You must specify either node[\'appdynamics\'][\'version\'] or node[\'appdynamics\'][\'dotnet_agent\'][\'version\']' unless version

package_source = agent['source']
unless package_source
  package_source = "#{node['appdynamics']['packages_site']}/java/#{version}/AppServerAgent-"
  package_source << 'ibm-' if agent['ibm_jvm']
  package_source << "#{version}.zip"
end

remote_file node['appdynamics']['java_agent']['zip'] do
  source package_source
  checksum agent['checksum']
  backup false
  mode '0444'
end

ark 'javaagent' do
  url "file:///#{node['appdynamics']['java_agent']['zip']}"
  path agent['install_dir']
  owner agent['owner']
  action :put
end

directory "#{agent['install_dir']}/javaagent/conf" do
  owner agent['owner']
  group agent['group']
  mode '0755'
end

directory "#{agent['install_dir']}/javaagent/logs" do
  owner agent['owner']
  group agent['group']
  mode '0755'
end

template "#{agent['install_dir']}/javaagent/conf/controller-info.xml" do
  cookbook agent['template']['cookbook']
  source agent['template']['source']
  owner agent['owner']
  group agent['group']
  sensitive true
  mode '0600'

  variables(
    app_name: node['appdynamics']['app_name'],
    tier_name: node['appdynamics']['tier_name'],
    node_name: node['appdynamics']['node_name'],

    controller_host: controller['host'],
    controller_port: controller['port'],
    controller_ssl: controller['ssl'],
    controller_user: controller['user'],
    controller_accesskey: controller['accesskey']
  )
end


recipes/machine_agent.rb
agent = node['appdynamics']['machine_agent']
controller = node['appdynamics']['controller']
http_proxy = node['appdynamics']['http_proxy']

version = agent['version'] || node['appdynamics']['version']
raise 'You must specify either node[\'appdynamics\'][\'version\'] or node[\'appdynamics\'][\'machine_agent\'][\'version\']' unless version

agent_zip = "#{Chef::Config[:file_cache_path]}/AppDynamicsMachineAgent.zip"
package_source = agent['source']

unless package_source
  package_source = "#{node['appdynamics']['packages_site']}/machine/#{version}/"

  if agent['use_bundled_package']
    package_source << 'machineagent-bundle-'
    package_source << if node['kernel']['machine'] == 'x86_64'
                        '64bit-'
                      else
                        '32bit-'
                      end

    package_source << value_for_platform_family(
      %w(mswin windows) => 'windows',
      %w(darwin mac_os_x) => 'osx',
      %w(omnios opensolaris solaris solaris2 smartos) => 'solaris',
      'default' => 'linux'
    )
  else
    package_source << 'MachineAgent'
  end

  package_source << "-#{version}.zip"
end

package 'unzip'

directory "#{agent['install_dir']}/conf" do
  owner agent['owner']
  group agent['group']
  mode '0755'
  recursive true
  action :create
end

remote_file agent_zip do
  source package_source
  checksum agent['checksum']
  backup false
  mode '0444'
  notifies :run, 'execute[unzip-appdynamics-machine-agent]', :immediately
end unless ::File.exist?([agent['install_dir'], '/bin'].join)

template "#{agent['install_dir']}/run.sh" do
  cookbook agent['run_sh']['cookbook']
  source agent['run_sh']['source']
  owner agent['owner']
  group agent['group']
  mode '0744'
  variables(
    java: agent['java'],
    java_params: agent['java_params'],
    install_dir: agent['install_dir'],
    pid_file: agent['pid_file']
  )
end

execute 'unzip-appdynamics-machine-agent' do
  cwd agent['install_dir']
  command "unzip -qqo #{agent_zip}"
  action :nothing
end

template agent['init_script'] do
  cookbook agent['init']['cookbook']
  source agent['init']['source']
  variables(
    install_dir: agent['install_dir'],
    pid_file: agent['pid_file']
  )
  owner agent['owner']
  group agent['group']
  mode '0744'
end

template "#{agent['install_dir']}/conf/controller-info.xml" do
  cookbook agent['template']['cookbook']
  source agent['template']['source']
  owner agent['owner']
  group agent['group']
  sensitive true
  mode '0600'

  variables(
    app_name: node['appdynamics']['app_name'],
    tier_name: node['appdynamics']['tier_name'],
    node_name: node['appdynamics']['node_name'],
    unique_host_id_name: node['appdynamics']['unique_host_id_name'],

    controller_host: controller['host'],
    controller_port: controller['port'],
    controller_ssl: controller['ssl'],
    controller_user: controller['user'],
    controller_accesskey: controller['accesskey'],

    http_proxy_host: http_proxy['host'],
    http_proxy_port: http_proxy['port'],
    http_proxy_user: http_proxy['user'],
    http_proxy_password_file: http_proxy['password_file']
  )
  notifies :restart, 'service[appdynamics_machine_agent]', :delayed
end

service 'appdynamics_machine_agent' do
  supports [:start, :stop, :restart]
  action [:enable, :start]
end
